### PR TITLE
Wojownicy kodu us2

### PR DESCRIPTION
**Pull request by adamvm**

### CODE REVIEW

File: src/CourseDashboard.cpp [hunk 1]
```diff
@@ -51,34 +54,88 @@ void CourseDashboard::updateUser(User & user)
...
+            break;
+    }
+}
+
+void CourseDashboard::saveToFile()
+{
+    std::fstream file_DB("baza danych.txt", file_DB.in | file_DB.out);
+    if(!file_DB.is_open())
+    {
+        std::cout << "error" << std::endl;
```
Comment (ziobron) [inline src/CourseDashboard.cpp:97]: how to test if a file was not open? This implementation is difficult to test. It's better to throw an exception.

File: src/CourseDashboard.cpp [hunk 2]
```diff
@@ -51,34 +54,88 @@ void CourseDashboard::updateUser(User & user)
...
+    }
+    else
+    {    
+        for (auto & user : users_)
+        {
+            file_DB << user.getAllInfo();
+        }
+    }
+    file_DB.close();
```
Comment (ziobron) [inline src/CourseDashboard.cpp]: You don't need to close. fstream implements a RAII, so it closes file automatically when the destructor is called.

File: src/CourseDashboard.cpp [hunk 3]
```diff
@@ -51,34 +54,88 @@ void CourseDashboard::updateUser(User & user)
...
+    std::fstream file_DB("baza danych.txt", file_DB.in);
+    if(!file_DB.is_open())
+    {
+        std::cout<<"No file userBase.txt found."<<std::endl;
+    }
+    else
+    {    
+        std::string line;
+        std::vector<std::string> usersData{};
+	    std::vector<std::string> user{};
```
Comment (ziobron) [inline src/CourseDashboard.cpp:119]: formatting - indentation

File: src/CourseDashboard.cpp [hunk 4]
```diff
@@ -51,34 +54,88 @@ void CourseDashboard::updateUser(User & user)
...
+            }
+            size_t length = element.find("|");
+            user.push_back(element.substr(0,length));
+
+            User newUser(user[0], user[1], user[2], user[3], user[4]);
+            createUser(newUser);
+            user.clear();
+        }	
     }
+    file_DB.close();
```
Comment (ziobron) [inline src/CourseDashboard.cpp]: no need to close, as above

File: src/CourseDashboard.cpp [hunk 5]
```diff
@@ -51,34 +54,88 @@ void CourseDashboard::updateUser(User & user)
...
+        {
+            file_DB << user.getAllInfo();
+        }
+    }
+    file_DB.close();
+}
+
+void CourseDashboard::loadFromFile()
+{
+    std::fstream file_DB("baza danych.txt", file_DB.in);
```
Comment (ziobron) [inline src/CourseDashboard.cpp]: maybe the file name should be a parameter to this function? You can use ifstream and skip the in flag.

File: src/CourseDashboard.cpp [hunk 6]
```diff
@@ -51,34 +54,88 @@ void CourseDashboard::updateUser(User & user)
...
+            break;
+        default:
+            std::cout << "You gave wrong value.";
+            break;
+    }
+}
+
+void CourseDashboard::saveToFile()
+{
+    std::fstream file_DB("baza danych.txt", file_DB.in | file_DB.out);
```
Comment (ziobron) [inline src/CourseDashboard.cpp]: the filename could be a parameter. BTW ofstream would be enough here, so you would not need to provide flags in and out. Only out should be used here.

File: src/User.hpp [hunk 7]
```diff
@@ -6,6 +6,8 @@
 class User
 {
 public:
+    User() = default;
```
Comment (ziobron) [inline src/User.hpp]: we don't want to have this construtor. Please remove it.

File: tests/CourseDashboardTests.cpp [hunk 8]
```diff
@@ -1,8 +1,29 @@
...
+                         "jkowalski");
+    }
+};
 
 
-TEST(CourseDashboardTests, canConstructCourseDashboard)
+TEST_F(CourseDashboardTests, canConstructCourseDashboard)
 {
     CourseDashboard cd;
 }
```
Comment (ziobron) [inline tests/CourseDashboardTests.cpp:37]: Looks like not tested :( You must provide unit tests.

**Review by ziobron — CHANGES_REQUESTED** [review]

File: src/CourseDashboard.cpp (hunk 1 above)
Comment (adamvm) [reply src/CourseDashboard.cpp:97]: changed to throw

File: src/CourseDashboard.cpp (hunk 2 above)
Comment (adamvm) [reply src/CourseDashboard.cpp]: Done

File: src/CourseDashboard.cpp (hunk 4 above)
Comment (adamvm) [reply src/CourseDashboard.cpp]: Done

File: src/User.hpp (hunk 7 above)
Comment (adamvm) [reply src/User.hpp]: Done

File: tests/CourseDashboardTests.cpp (hunk 8 above)
Comment (adamvm) [reply tests/CourseDashboardTests.cpp:37]: In progress

Comment (adamvm) [reply tests/CourseDashboardTests.cpp:37]: Current approach partial implementation in https://github.com/coders-school/CourseDashboard/pull/14/commits/53fed699000193c9997dba86926c56c3df846360

File: src/CourseDashboard.cpp (hunk 1 above)
Comment (adamvm) [reply src/CourseDashboard.cpp:97]: Problem! Found that was not changed in the commits and local copy - probably my mistake during marking as resolved - will be fixed

File: src/CourseDashboard.cpp [hunk 9]
```diff
@@ -105,9 +105,9 @@ void CourseDashboard::saveToFile()
...
 }
 
-void CourseDashboard::loadFromFile()
+void CourseDashboard::loadFromFile(std::string fileName)
 {
-    std::fstream file_DB("baza danych.txt", file_DB.in);
+    std::ifstream file_DB(fileName);
     if(!file_DB.is_open())
     {
         throw "Problem to open userBase.txt";
```
Comment (adamvm) [inline src/CourseDashboard.cpp:113]: Issue - fixed filename in the error message - can be deleted or better string concatenation

**Review by adamvm — COMMENTED** [review]
@MonikaZelechowska , I found some issues that can be improved

File: src/CourseDashboard.cpp [hunk 10]
```diff
@@ -89,9 +89,9 @@ void CourseDashboard::updateUser(User & user)
     }
 }
 
-void CourseDashboard::saveToFile()
+void CourseDashboard::saveToFile(std::string fileName)
 {
```
Comment (adamvm) [inline src/CourseDashboard.cpp:93]: Was it tested that filename with space works? If not we need include it in tests

File: src/CourseDashboard.cpp (hunk 9 above)
Comment (MonikaGajda) [reply src/CourseDashboard.cpp:113]: About function "save To File" I have now realized that checking if file is open is not necessary. Always new file will be created or replaced if there exist one with the same name. If I'm not right, please correct me.

Comment (adamvm) [reply src/CourseDashboard.cpp:113]: Old one can be openned in other program e.g. notepad++ and overwriting will be not possible, I think

File: src/CourseDashboard.cpp (hunk 10 above)
Comment (MonikaGajda) [reply src/CourseDashboard.cpp:93]: Yes, filename with space works

File: src/User.hpp [hunk 11]
```diff
@@ -6,13 +6,14 @@
...
-    User(const std::string & name,
-         const std::string & nick,
-         const std::string & group,
-         const std::string & gitHub,
-         const std::string & firecode);
+    User(const std::string name,
+         const std::string nick,
+         const std::string group,
+         const std::string gitHub,
+         const std::string firecode);
```
Comment (ziobron) [inline src/User.hpp:13]: why not references? std::string_view can be used as well

File: src/User.cpp [hunk 12]
```diff
@@ -15,15 +15,14 @@ User::User(const std::string & name,
...
-    ss<<std::setw(1) << "5." << firecode_ << "|";
-    ss<<std::endl;
-    return ss.str();
+    std::string output;
+    output.append("1.").append(name_).append("|");
+    output.append("2.").append(nick_).append("|");
+    output.append("3.").append(group_).append("|");
+    output.append("4.").append(gitHub_).append("|");
+    output.append("5.").append(firecode_).append("|");
+    output.append("\n");
```
Comment (ziobron) [inline src/User.cpp:24]: This approach causes a lot of additional memory allocations. It's better to use stringstream for such concatenations as it was used before.

File: src/main.cpp [hunk 13]
```diff
@@ -6,12 +6,24 @@ int main()
...
     cdb.createUser(u1);
-    cdb.retriveUserByNick("SzymonGajewski");
-    cdb.showAll();
-    cdb.deleteUserByNick("SzymonGajewski");
-    std::cout<<"after delete: "<<std::endl;
-    cdb.showAll();
+
+    std::cout << cdb.showAll();
+
+    cdb.retriveUserByNick("PiotrNowak");  // TODO improve function
```
Comment (ziobron) [inline src/main.cpp:22]: yup, returning the user would be much better :)

File: tests/CourseDashboardTests.cpp [hunk 14]
```diff
@@ -1,8 +1,53 @@
...
+                  "jkowalski",
+                  "jkowalski"};
+
+    User testUser2{"Stefan",
+                  "StefanNowak",
+                  "weekendowa",
+                  "snowak",
+                  "snowak"};
+
+    CourseDashboardTests() = default;
```
Comment (ziobron) [inline tests/CourseDashboardTests.cpp:22]: not needed - to be removed

File: tests/CourseDashboardTests.cpp [hunk 15]
```diff
@@ -1,8 +1,53 @@
...
+
+    void SetUp( ) override {
+        CourseDashboard_.createUser(testUser1);
+    }
+
+    void TearDown() override {
+
+    }
```
Comment (ziobron) [inline tests/CourseDashboardTests.cpp:30]: empty function not needed - to be removed

File: tests/CourseDashboardTests.cpp [hunk 16]
```diff
@@ -1,8 +1,53 @@
...
+    int recordsCount = std::count(result.begin(), result.end(), '\n');
+    EXPECT_EQ(recordsCount, 2);
+}
+
+TEST_F(CourseDashboardTests, canDeleteUser)
+{
+    CourseDashboard_.deleteUserByNick("JanKowalski");
+    auto result = CourseDashboard_.showAll();
+    EXPECT_EQ(result, "");
+}
```
Comment (ziobron) [inline tests/CourseDashboardTests.cpp:52]: Here are test for the US1 functionality (it's ok). 
But there also should be tests for US2.

File: src/CourseDashboard.cpp [hunk 17]
```diff
@@ -2,23 +2,28 @@
 #include <stdexcept>
 #include <iostream>
 #include <algorithm>
+#include <fstream>
+#include <string>
 
-CourseDashboard::CourseDashboard() {}
 
-void CourseDashboard::showAll()
+CourseDashboard::CourseDashboard() = default;
```
Comment (ziobron) [inline src/CourseDashboard.cpp:9]: constructor not needed. We don't have another so it will be generated anyway.

File: src/CourseDashboard.cpp [hunk 18]
```diff
@@ -2,23 +2,28 @@
...
+
+std::string CourseDashboard::showAll()
 {
+    std::string result;
     for (auto & user : users_)
     {
-        std::cout << user.getAllInfo();
+        result.append(user.getAllInfo());
```
Comment (ziobron) [inline src/CourseDashboard.cpp:16]: ostringstream could be a bit more efficient instead of append

File: src/CourseDashboard.cpp [hunk 19]
```diff
@@ -51,34 +56,86 @@ void CourseDashboard::updateUser(User & user)
...
+            break;
+        default:
+            std::cout << "You gave wrong value.";
+            break;
+    }
+}
+
+void CourseDashboard::saveToFile(std::string fileName)
+{
+    std::ofstream file_DB(fileName);
```
Comment (ziobron) [inline src/CourseDashboard.cpp:94]: file naming conventions - snake case is here, but we use camelCase in this project.

File: src/CourseDashboard.cpp (hunk 9 above)
Comment (ziobron) [reply src/CourseDashboard.cpp:113]: Rather not. If the file is opened in another program overwriting will still be possible. But we may not have permissions to open the file if someone has changed it.

**Review by ziobron — CHANGES_REQUESTED** [review]
Still missing tests

**Comment by ziobron** [comment]
+7 points (lack of tests)

File: src/User.hpp (hunk 11 above)
Comment (adamvm) [reply src/User.hpp:13]: My mistake

File: src/User.cpp (hunk 12 above)
Comment (adamvm) [reply src/User.cpp:24]: My mistake, code readibility was horrible

File: src/CourseDashboard.cpp (hunk 18 above)
Comment (adamvm) [reply src/CourseDashboard.cpp:16]: Python way :/ :)

**Pull request closed without merging**